core/histogram.cpp

```diff
@@ -6,71 +6,128 @@
+      if (weights_unit != units::counts)
+        throw except::UnitError(
+            "Weights of sparse data must be `units::counts`.");
+      return weights_unit;
+    };
+
 DataArray histogram(const DataConstProxy &sparse,
                     const VariableConstProxy &binEdges) {
-  if (sparse.hasData())
-    throw except::SparseDataError(
-        "`histogram` is not implemented for sparse data with values yet.");
-  if (sparse.dims().ndim() > 1)
-    throw std::logic_error("Only the simple case histograms may be constructed "
-                           "for now: 2 dims including sparse.");
   auto dim = binEdges.dims().inner();
-  if (binEdges.unit() != sparse.coords()[dim].unit())
-    throw std::logic_error(
-        "Bin edges must have same unit as the sparse input coordinate.");
-  if (binEdges.dtype() != dtype<double> ||
-      sparse.coords()[dim].dtype() != DType::Double)
-    throw std::logic_error("Histogram is only available for double type.");
 
   auto result = apply_and_drop_dim(
       sparse,
       [](const DataConstProxy &sparse_, const Dim dim_,
          const VariableConstProxy &binEdges_) {
-        const auto &coord = sparse_.coords()[dim_].sparseValues<double>();
-        auto edgesSpan = binEdges_.values<double>();
-        expect::histogram::sorted_edges(edgesSpan);
-        // Copy to avoid slow iterator obtained from VariableConstProxy.
-        const std::vector<double> edges(edgesSpan.begin(), edgesSpan.end());
-        auto resDims{sparse_.dims()};
-        auto len = binEdges_.dims()[dim_] - 1;
-        resDims.resize(resDims.index(dim_), len);
-        Variable res =
-            makeVariableWithVariances<double>(resDims, units::counts);
-        if (scipp::numeric::is_linspace(edges)) {
-          // Special implementation for linear bins. Gives a 1x to 20x speedup
-          // for few and many events per histogram, respectively.
-          const double offset = edges.front();
-          const double nbin = static_cast<double>(len);
-          const double scale = nbin / (edges.back() - edges.front());
-          for (scipp::index i = 0; i < sparse_.dims().volume(); ++i) {
-            auto curRes = res.values<double>().begin() + i * len;
-            for (const auto &c : coord[i]) {
-              const double bin = (c - offset) * scale;
-              if (bin >= 0.0 && bin < nbin)
-                ++(*(curRes + static_cast<scipp::index>(bin)));
-            }
-          }
+        if (sparse_.hasData()) {
+          return transform_subspan<std::tuple<
+              std::tuple<span<double>, sparse_container<double>,
+                         sparse_container<double>, span<const double>>,
+              std::tuple<span<double>, sparse_container<float>,
+                         sparse_container<double>, span<const double>>,
+              std::tuple<span<double>, sparse_container<float>,
+                         sparse_container<double>, span<const float>>>>(
```

Comment:
Note this ugly but still not exhaustive list. We may also need to add options that let us specify different output types.